### DISCUSSION TITLE
Random number generation

### DISCUSSION THREAD

**Discussion by ftessier** [darcymason/egsnrc#23] [General]
It occurred to me that if we are going to compare code efficiencies, we ought to get our random number generation straight; especially for photon simulations where it has a larger impact.

EGSnrc uses `ranmar` by default, and the mortran backend also has `ranlux`. Both of these are severely outdated. I propose that going forward we use `pcg64` or `xoshiro`, the latter being the [fastest according to its author](https://prng.di.unimi.it/). The `xoshiro256++` generator is faster than the old `ranmar` by a factor of $\sim$ 4.3 according to cursory testing and faster than the Mersenne Twister also by a factor of $\sim$ 4. The `xoshiro` generators have the distinct advantage of a jump ahead shortcut to generate independent rngs for parallel simulation.

In Rust, I can generate $\sim$ 1.25 billion random 64-bit numbers per second (64Gb/s) with `xoshiro256++` on my current workstation, compared to $\sim$ 0.29 64Gb/s for the egs++ implementation of ranmar. A straight implementation pf `xoshiro` in C using `gcc -O3` generates $\sim$ 0.88 64Gb/s. Presumably, `llvm` explains the advantage or Rust here, for such a small code. 

I tried the default numpy rng (pcg64) with standard python, but of course in vanilla python it is much slower, clocking in at $\sim$ 0.002 64Gb/s.

```python
#!/usr/bin/env python

import numba
import numpy
import random
import time

### seed generator
rng = numpy.random.default_rng(42)

### @numba.jit(nopython=True)
def generate_random_numbers():
    for i in range(int(1e9)):
        rng.random()
    return rng.random()

start_time = time.time()
result = generate_random_numbers()
end_time = time.time()

# Print the execution time
print(f"Final answer: number = {result}")
print(f"Execution time: {end_time - start_time:.2f} seconds")

```

No luck either trying to code `xoshiro` and compile it with numba. Anyways, that is moot because I can compile the short `xoshiro` C or Rust code as a library which is called from python.

@darcymason I see that your are setup with `numba.cuda` to get `xoshiro128+` random numbers. Can you provide a rough metric for generation speed compared to vanilla python `pcg64`, for a given GPU? I presume it is much much faster, and of course it then depends on the GPU.

| random number generator                               | relative generator speed |
| :---------------------------------------------------- | :----------------------- |
| numpy default (pcg64)                                 | 1.0                      |
| EGSnrc ranmar (C++, `g++ -O3`, `-legspp -lm`)         | 13.2                     |
| xoshiro256++ (C, `gcc -O3`)                           | 400                      |
| xoshiro256++ (Fortran90, `gfortran -O3`)              | ?                        |
| xoshiro256++ (Rust `--release`, `rustc`, i.e. `llvm`) | 568                      |
| xoshiro128+ (`numba.cuda`, GPU)                       | ?                        |

**Comment by darcymason** [comment]
> @darcymason I see that your are setup with numba.cuda to get xoshiro128+ random numbers. Can you provide metrics for generation speed compared to vanilla python pcg64, for a given GPU?

@ftessier, sure, I'll have a look in the next couple of days.  I did have it running already for some tests where I generated a spread of energies for the photon source. I can just isolate that for generating a comparison.

**Comment by darcymason** [comment]
@ftessier, in starting to write this for numba cuda, I realized that the jit might optimize away the random numbers generated if nothing is done with them afterwards.  Do you know whether that is a possibility in the Rust of C++ code you measured above?  I'm not familiar, with what, e.g. -O3 attempts to do.  I'm wondering whether there is a way to ensure that they are compared similarly, e.g. storing some kind of output.  In the Python example you gave above, I suppose a smart optimizer should know that even though the random numbers generated in the loop were never used, they would affect the last one that was returned.  But in GPU we don't have that kind of option, since each generation happens independently in its own thread.

**Reply by ftessier** [reply]
Very good point. That is why I print a random number from the sequence at the end, which ensures the compiler cannot optimize away the sampling loop. If it could then the rng would be a very poor one!! 🤓 Running with `perf` confirms that the number of instructions executed scales with $\large N$.

The Rust and C code give efficiencies that are on par with what is expected, ~1 billion random numbers per second (on my workstation which dates a few years). It is ~4 times faster than ranmar, which makes sense from the source codes for both.

For EGSnrc, we need to generate 64-bit random numbers, 32-bit leading to float grid issues in some applications, e.g., nanometer size geometries.

I recall that GPUs were geared toward 32-bit efficiency, but I presume that has evolved as well?

**Reply by darcymason** [reply]
64-bit is still much slower, I think I read 2-4 times slower for this kind of work but I'll have to check again.  I've been working with 32-bits, but the size is in a config file so can be easily modified by the user.

From early results, I think my randoms are not getting optimized away, but need to put it aside for today and come back to this tomorrow.

**Reply by ftessier** [reply]
Excellent, if it is a config option then no worries! I guess 2 times slower is as good as it gets; or did you mean 2 times slower _per bit?_ Anyways it doesn't matter, given the user option.

**Comment by darcymason** [comment]
Have some results, currently in #24.

Bottom line is (on Simon's machine on one GPU) seeing about 690 Billion numbers per second (0.725 s for 500B).  GPU utilization was 100% for this test.

For numba cuda's xoroshiro128p random generator, 64-bit tests were the same as 32-bit.  The [source code](https://github.com/numba/numba/blob/main/numba/cuda/random.py) shows they call the same function and just convert the result to 64 or 32 bits. 

I did store one random number per GPU block to try to ensure the calls weren't optimized away.  But if that is removed, the times are almost unchanged.

But in general, 64-bits will likely be a fair bit slower than 32.  In numba.cuda.detect() for Simon's machine, there is a statement, "FP32/FP64 Performance Ratio: 32".  But there is no effect on the random numbers with current numba.

**Reply by ftessier** [reply]
That is fantastic news, an impressive rng generation speed (and code writing speed!) All generators work with integers. To return floats the `uint` is simply divided by `MAX_UINT`. So any 32 to 64 bit _floating-point_ performance degradation on GPU is irrelevant.

**Reply by SimonBiggs** [reply]
Wow... so about a 500x improvement by using Python with a 3090 GPU over using Rust on the CPU?

**Reply by ftessier** [reply]
I was surprised by @darcymason's comment for 64 vs 32 bit efficiency comment. But since it converts from `uint64` to `uint32`, it makes sense. I guess it would be possible to extract two 32-bit `uints` then, but the low bits would not have the same quality, and generally I would not toy with a well-crafted rng. History shows that messing with rng's can severely degrade their quality, e.g., the `RANDU` generator!! Anyways at 500e9/s, this is perhaps not going to be the bottleneck in EGSnrc. 😄

**Reply by darcymason** [reply]
> Wow... so about a 500x improvement by using Python with a 3090 GPU over using Rust on the CPU?

It seems that way - although this is so sensitive to the measuring setup.  Of course it also depends on the ~100% CPU that this simple loop can achieve.  In real use, mingled in with other operations, there will of course be some drop.

The other thing I forgot to mention is that this timing does not include an initial creation of the `rng_states` - setting up the parallel threads so they each have their own separated space of random numbers.  There were num_blocks * threads_per_block states set up, which typically were each 512 or 1024 in these tests, so let's say 1M states, vs 500B in the run, so almost certainly negligible, but I don't now how much computing goes into that initialization.   In any case, it is a fixed initial overhead, you could run 500T and it would still be the same.

At some point it would be good to have a test to ensure this is being used properly and that randomization tests pass - I wouldn't be overly surprised if I somehow set up the states incorrectly and distinct random space was not actually being used properly.

**Reply by ftessier** [reply]
I recommend testing with "BigCrush": https://github.com/lemire/testingRNG

**Reply by darcymason** [reply]
Numba docs [claim](https://numba.readthedocs.io/en/stable/cuda/random.html) that xoroshiro128+ passes BigCrush (I assume/hope they mean their implementation of it?) but perhaps best to do something ourselves, too. I'm going to set it aside for now, perhaps eventually someone with better knowledge than me can try to tackle that once the code is filled in some more.

**Reply by ftessier** [reply]
Yes, I think this is very low priority. The xoshiro generators are widespread (default in an increasing number of libraries) and heavily tested, so I would for one be confident the numba implementers have tested _their_ implementation of it (although indeed that is not clear).

Anyways, the _randomness_ quality of the rng is not very important in radiation Conte Carlo, because the simulation itself is 'stochastic', so to speak, in the sense that random numbers are not used for the same 'thing' from one history to the next. I mean, we do often model a monoenergetic pencil beam source, which is 100% non-random! :-)
